lib/cretonne/meta/cdsl/typevar.py

```diff
@@ -241,6 +263,109 @@ def __iand__(self, other):
 
         return self
 
+    def lane_of(self):
+        # type: () -> TypeSet
+        """
+        Return a TypeSet describing the image of self across lane_of
+        """
+        new = self.copy()
+        new.lanes = set([1])
+        return new
+
+    def as_bool(self):
+        # type: () -> TypeSet
+        """
+        Return a TypeSet describing the image of self across as_bool
+        """
+        new = self.copy()
+        new.ints = set()
+        new.floats = set()
+        new.bools = self.ints.union(self.floats).union(self.bools)
+
+        if 1 in self.lanes:
+            new.bools.add(1)
+        return new
+
+    def half_width(self):
+        # type: () -> TypeSet
+        """
+        Return a TypeSet describing the image of self across halfwidth
+        """
+        new = self.copy()
+        new.ints = set([x/2 for x in self.ints if x > 8])
+        new.floats = set([x/2 for x in self.floats if x > 32])
+        new.bools = set([x/2 for x in self.bools if x > 8])
+
+        return new
+
+    def double_width(self):
+        # type: () -> TypeSet
+        """
+        Return a TypeSet describing the image of self across doublewidth
+        """
+        new = self.copy()
+        new.ints = set([x*2 for x in self.ints if x < MAX_BITS])
+        new.floats = set([x*2 for x in self.floats if x < MAX_BITS])
+        new.bools = set(filter(legal_bool,
+                               set([x*2 for x in self.bools if x < MAX_BITS])))
+
+        return new
+
+    def half_vector(self):
+        # type: () -> TypeSet
+        """
+        Return a TypeSet describing the image of self across halfvector
+        """
+        new = self.copy()
+        new.lanes = set([x/2 for x in self.lanes if x > 1])
+
+        return new
+
+    def double_vector(self):
+        # type: () -> TypeSet
+        """
+        Return a TypeSet describing the image of self across doublevector
+        """
+        new = self.copy()
+        new.lanes = set([x*2 for x in self.lanes if x < MAX_LANES])
+
+        return new
+
+    def size(self):
+        # type: () -> int
+        """
+        Return the number of concrete types represented by this typeset
+        """
+        return len(self.lanes) * (len(self.ints) + len(self.floats) +
+                                  len(self.bools))
+
+    def get_singleton(self):
+        # type: () -> types.ValueType
+        """
+        Return the singleton type represented by self. Can only call on
+        typesets containing 1 type.
+        """
+        assert self.size() == 1
+        if len(self.ints) > 0:
+            bits = tuple(self.ints)[0]
+            scalar_type = ValueType.by_name(IntType.get_name(bits))
```

Comment:
If `IntType` had a registry of created integer types and a static `with_bits()` method, you wouldn't need this stringly typed stuff. I think mypy would be happier too.